### PR TITLE
Replace caching engine

### PR DESCRIPTION
**Pull request by aerosol**
### Changes

This PR swaps out [Cachex](https://github.com/whitfin/cachex) for [ConCache](https://github.com/sasa1977/con_cache). The former feels too advanced for our needs, plus with https://github.com/plausible/analytics/commit/83a46fb47e96f51f058ad7d3315b0d218c331c6c we're not evicting user agents properly (ref: https://github.com/whitfin/cachex/issues/329).

All application caches are now accessed through a single interface (`Plausible.Cache.Adapter`), even if they don't implement the periodic refresh semantics. This lets us test drive ConCache uniformly and offers future-proof flexibility.

Notable mentions:

- ConCache alone does not offer hit/miss ratio tracking, so this was added so we can still track utilization in grafana. Metric names will change (`s/cachex/cache`), we could stick with keeping the `cachex` string, but I was annoyed by that.
- ConCache does not offer cache eviction by specifying the number of items, but this isn't a hard requirement - from now on, the User Agents cache uses (roughly) 1 hour global TTL per item
- merging periodic cache refreshes is now much faster, since we're levering dirty ets writes 
- all caches are now configured at the supervision tree definition level, arguably resulting in a better overview

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [x] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [ ] The UI has been tested both in dark and light mode
- [x] This PR does not change the UI

### CODE REVIEW

File: lib/plausible/telemetry/plausible_metrics.ex [hunk 1]
```diff
@@ -88,29 +88,23 @@ defmodule Plausible.PromEx.Plugins.PlausibleMetrics do
...
-      hit_rate: user_agents_hit_rate
+    :telemetry.execute([:prom_ex, :plugin, :cache, :user_agents], %{
+      count: user_agents_stats.count,
+      hit_rate: user_agents_stats.hit_rate
     })
 
-    :telemetry.execute([:prom_ex, :plugin, :cachex, :sessions], %{
-      count: sessions_count,
-      hit_rate: sessions_hit_rate
+    :telemetry.execute([:prom_ex, :plugin, :cache, :sessions], %{
```
Comment (zoldar) [inline lib/plausible/telemetry/plausible_metrics.ex:102]: Just a note that we'll need to update respective graphs/dashboards in Grafana, right?

**Review by zoldar — APPROVED** [review]

File: lib/plausible/telemetry/plausible_metrics.ex (hunk 1 above)
Comment (aerosol) [reply lib/plausible/telemetry/plausible_metrics.ex:102]: yup

File: lib/plausible/cache/adapter.ex [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  @spec size(atom()) :: non_neg_integer() | nil
+  def size(cache_name) do
+    try do
+      ConCache.size(cache_name)
+    catch
+      :exit, _ -> nil
+    end
+  end
```
[the remark this answers is not in the document]
Comment (aerosol) [reply lib/plausible/cache/adapter.ex]: Ah excellent. bac7fde2f

File: lib/plausible/cache.ex [hunk 3]
```diff
@@ -165,21 +165,8 @@ defmodule Plausible.Cache do
...
-      end
-
-      @spec hit_rate() :: number()
-      def hit_rate(cache_name \\ name()) do
-        case Cachex.stats(cache_name) do
-          {:ok, stats} -> Map.get(stats, :hit_rate, 0)
-          _ -> 0
-        end
-      end
+      defdelegate size(cache_name \\ name()), to: Plausible.Cache.Stats
```
[the remark this answers is not in the document]
Comment (aerosol) [reply lib/plausible/cache.ex]: d6717b779

**Comment by cnkk** [comment]
> ConCache does not offer cache eviction by specifying the number of items, but this isn't a hard requirement - from now on, the User Agents cache uses (roughly) 1 hour global TTL per item

If increasing this helps with performance, let's do it?

**Comment by aerosol** [comment]
> > ConCache does not offer cache eviction by specifying the number of items, but this isn't a hard requirement - from now on, the User Agents cache uses (roughly) 1 hour global TTL per item
> 
> If increasing this helps with performance, let's do it?

No proof it does - first cold UA takes a bit longer (even up to 10ms, YMMV), but how many of them come back frequently enough for this to be effective is unknown. I think we must be wary of modest setups, memory-wise, running this too. Perhaps I should make this configurable via env?

**Comment by aerosol** [comment]
> even up to 10ms

up to 100ms

**Review by ruslandoga — APPROVED** [review]

File: lib/plausible/telemetry/plausible_metrics.ex (hunk 1 above)
Comment (aerosol) [reply lib/plausible/telemetry/plausible_metrics.ex:102]: in the end - no - it's only the event shape that changes, metric names remain the same 😅

**Pull request merged**